### ISSUE TITLE
ArrayIndexOutOfBoundsException

### ISSUE THREAD

**Issue by zhouchaoyuan** [zxing/zxing#862]
java.lang.ArrayIndexOutOfBoundsException: length=929; index=-158
at com.google.zxing.pdf417.decoder.ec.ModulusGF.multiply(ModulusGF.java:105)
at com.google.zxing.pdf417.decoder.ec.ModulusPoly.evaluateAt(ModulusPoly.java:99)
at com.google.zxing.pdf417.decoder.ec.ErrorCorrection.decode(ErrorCorrection.java:53)
at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.correctErrors(PDF417ScanningDecoder.java:553)
at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.decodeCodewords(PDF417ScanningDecoder.java:526)
at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.createDecoderResultFromAmbiguousValues(PDF417ScanningDecoder.java:307)
at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.createDecoderResult(PDF417ScanningDecoder.java:276)
at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.decode(PDF417ScanningDecoder.java:123)
at com.google.zxing.pdf417.PDF417Reader.decode(PDF417Reader.java:87)
at com.google.zxing.pdf417.PDF417Reader.decode(PDF417Reader.java:61)
at com.google.zxing.MultiFormatReader.decodeInternal(MultiFormatReader.java:171)
at com.google.zxing.MultiFormatReader.decodeWithState(MultiFormatReader.java:85)

I have found a problem above. In the file 
https://github.com/zxing/zxing/blob/7df4aa6694a547098fe1e3ab430e5095b93b7827/core/src/main/java/com/google/zxing/pdf417/decoder/ec/ModulusGF.java, the code in line 105 report ArrayIndexOutOfBoundsException,is it a problem?

**Comment by srowen** [comment]
it's a problem. I've never gotten a reproduction though. Do you have one?

**Comment by zhouchaoyuan** [comment]
This issue was reported by automated testing.  I never reproduct it too.
Can we try 
`expTable[((logTable[a] + logTable[b]) % (modulus - 1) + (modulus - 1)) % (modulus - 1)];`
to solve the problem?

**Comment by srowen** [comment]
No, that can't be the problem. logTable and expTable can't have negative values to begin with. The error must be that one of the args is negative, and that also shouldn't happen. Any more error checking here doesn't change the issue. It really needs a reproduction.

**Comment by laeubi** [comment]
Maybe it would help here to replace this
`return expTable[(logTable[a] + logTable[b]) % (modulus - 1)];`
with
```
    int index = logTable[a] + logTable[b] % (modulus - 1);
    if (index >= expTable.length || index < 0) {
    	throw new IndexOutOfBoundsException("index "+index+" is out of bounds (modulus="+modulus+", logTable["+a+"]="+logTable[a]+", logTable["+b+"]="+logTable[b]+")");
    }
	return expTable[index];
```
That way we know the args and calculated values and one can guess whats wrong then or how to reproduce it?

**Comment by srowen** [comment]
Yeah that's about the only way I can think to debug this. Maybe even print the whole array in this case

**Comment by lueis** [comment]
I have the same problem in cpp. You should check for a and b. I will test this with:
`a&=0xFF;b&=0xFF` but its very static.

**Comment by rmunn** [comment]
Here's a QR code that just triggered an ArrayIndexOutOfBoundsException when I submitted it to https://zxing.org/w/decode.jspx — I believe this might be a reproduction of this issue:

![qr-code-large](https://user-images.githubusercontent.com/90762/37891775-b88523fc-30ff-11e8-85f7-2d5fa1cab1fb.gif)

The contents of this QR are probably MIT-licensed (and not copyright by me) since I created it by going to https://tiddlywiki.com/prerelease/, clicking the QR code button, and choosing the "QR code of raw tiddler content" option. This is the text that a different decoder (http://blog.qr4.nl/Online-QR-Code_Decoder.aspx) successfully decoded from that QR code:

```plaintext
This is a pre-release build of TiddlyWiki, [[also available in empty form|https://tiddlywiki.com/prerelease/empty.html]]. It is provided for testing purposes. ''Please don't try to use it for anything important'' -- you should use the latest official release from https://tiddlywiki.com.

<$list filter="[tag[ReleaseNotes]!has[released]!sort[created]]">
<h1><$text text=<<currentTiddler>>/></h1>
<$transclude mode="block"/>
</$list>

! Features held back for later release

These unfinished features are included in this prerelease build, but are not planned for inclusion in the next release.

* <$button>
<$action-setfield $tiddler="$:/view" text="stacked"/>
Set "Stacked" story view
</$button> <$button>
<$action-setfield $tiddler="$:/view" text="classic"/>
Revert
</$button>
```

**Comment by srowen** [comment]
Ah, this seems to trigger an error in the image reading library, ImageIO:

```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 4096
	at com.sun.imageio.plugins.gif.GIFImageReader.read(GIFImageReader.java:984)
	at javax.imageio.ImageIO.read(ImageIO.java:1448)
	at javax.imageio.ImageIO.read(ImageIO.java:1400)
	at com.google.zxing.client.j2se.ImageReader.readImage(ImageReader.java:44)
	at com.google.zxing.client.j2se.DecodeWorker.decode(DecodeWorker.java:120)
	at com.google.zxing.client.j2se.DecodeWorker.call(DecodeWorker.java:76)
	at com.google.zxing.client.j2se.CommandLineRunner.main(CommandLineRunner.java:99)
```

If you save it as a PNG, you get the correct decoded text: "This is a pre-release build of TiddlyWiki ..."

This isn't actually a reproduction of the bug I'm afraid.

**Comment by mw0516** [comment]
Has this bug been fixed

**Comment by madisp** [comment]
Hi, I've seen this in the wild and managed to reproduce via fuzzing with JQF+AFL.

Input file:

![aioob](https://user-images.githubusercontent.com/1147723/83650255-cffd2880-a5c0-11ea-950d-4e68ffb98d61.png)

Reproducing code:

```java
File file = new File("/Users/madis/Pictures/aioob.png");
BufferedImage img = ImageIO.read(file);

int w = img.getWidth();
int h = img.getHeight();

byte[] lum = new byte[w * h];
for (int x = 0; x < w; x++) {
  for (int y = 0; y < h; y++) {
    // picture is either black or white pixels, read just the blue channel
    lum[y*w + x] = (byte)(img.getRGB(x, y) & 0xff);
  }
}

LuminanceSource source = new PlanarYUVLuminanceSource(lum, w, h, 0, 0, w, h, false);
Binarizer binarizer = new HybridBinarizer(source);
BinaryBitmap bitmap = new BinaryBitmap(binarizer);
Result result = new PDF417Reader().decode(bitmap);
```

Stacktrace:

```
	at com.google.zxing.pdf417.decoder.ec.ModulusGF.multiply(ModulusGF.java:105)
	at com.google.zxing.pdf417.decoder.ec.ModulusPoly.evaluateAt(ModulusPoly.java:98)
	at com.google.zxing.pdf417.decoder.ec.ErrorCorrection.decode(ErrorCorrection.java:53)
	at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.correctErrors(PDF417ScanningDecoder.java:557)
	at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.decodeCodewords(PDF417ScanningDecoder.java:530)
	at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.createDecoderResultFromAmbiguousValues(PDF417ScanningDecoder.java:309)
	at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.createDecoderResult(PDF417ScanningDecoder.java:278)
	at com.google.zxing.pdf417.decoder.PDF417ScanningDecoder.decode(PDF417ScanningDecoder.java:124)
	at com.google.zxing.pdf417.PDF417Reader.decode(PDF417Reader.java:89)
	at com.google.zxing.pdf417.PDF417Reader.decode(PDF417Reader.java:63)
	at com.google.zxing.pdf417.PDF417Reader.decode(PDF417Reader.java:57)
```

**Comment by srowen** [comment]
Oh that's great, a reproduction. I'll look into it now to figure out how it happens if I can.

**Comment by srowen** [comment]
Should be resolved in https://github.com/zxing/zxing/commit/c89470b2b03cd953b71e5417a27b7ae16ea54e5b , thanks!

**Comment by ndreisg** [comment]
Thanks a lot for fixing this @srowen 
When can we expect the 3.4.1 version to be released?

**Comment by srowen** [comment]
Yeah time for a new release. I'll look into that soon.

**Comment by srowen** [comment]
I just released 3.4.1, FWIW.
